libraries/AP_HAL_ChibiOS/Scheduler.cpp

```diff
@@ -310,12 +324,50 @@ void Scheduler::_timer_thread(void *arg)
         if (sched->expect_delay_start != 0) {
             uint32_t now = AP_HAL::millis();
             if (now - sched->expect_delay_start <= sched->expect_delay_length) {
-                stm32_watchdog_pat();
+                sched->watchdog_pat();
             }
         }
     }
 }
 
+void Scheduler::_monitor_thread(void *arg)
+{
+    Scheduler *sched = (Scheduler *)arg;
+    chRegSetThreadName("apm_monitor");
+
+    while (!sched->_initialized) {
+        sched->delay(100);
+    }
+    bool using_watchdog = AP_BoardConfig::watchdog_enabled();
+
+    while (true) {
+        sched->delay(100);
+        if (using_watchdog) {
+            stm32_watchdog_save((uint32_t *)&hal.util->persistent_data, (sizeof(hal.util->persistent_data)+3)/4);
```

Comment:
There's no locking done on persistent_data, and we do have non-atomic values in there.  This is probably worth a warning and explanation of why inconsistent data is a far preferable alternative to mutexing.  Perhaps a comment on the definition of the pd structure?

Reply:
which non-atomic values are you referring to?